File: JAVA/ATM_Design/src/main/java/org/ceo/states/ReadCashWithdrawalDetailsState.java
package org.ceo.states;

import org.ceo.enums.ATMState;
import org.ceo.factories.CardManagerFactory;
import org.ceo.models.ATM;
import org.ceo.models.Card;
import org.ceo.services.CardManagerService;

public class ReadCashWithdrawalDetailsState extends ATMStateMachine {

    private final ATM atm;

    public ReadCashWithdrawalDetailsState(ATM atm) {
        this.atm = atm;
    }

    @Override
    public boolean cancelTransaction(int transactionId) {
        try {
            this.atm.changeState(new EjectingCardState(this.atm));
            return true;
        } catch (Exception e) {
            throw new IllegalStateException("Cannot Eject Card");
        }
    }

    @Override
    public boolean readCashWithdrawDetails(Card card, double amount, int transactionId) {
        CardManagerService manager = CardManagerFactory.getCardManager(card.getType());
        boolean isWithdrawValid = manager.validateWithdrawal(transactionId, amount);

        if(isWithdrawValid) {
            this.atm.changeState(new DispensingCashState(this.atm));
        } else {
            this.atm.changeState(new EjectingCardState(this.atm));
        }

        return isWithdrawValid;
    }

    @Override
    public ATMState getState() {
        return ATMState.READ_CASH_WITHDRAW_DETAILS;
    }
}

File: C++/ATM-System/src/states/ReadCashWithdrawalDetailsState.cpp
#include "../states/ReadCashWithdrawalDetailsState.h"
#include "../states/DispensingCashState.h"
#include "../models/ATM.h"
#include "../states/EjectingCardState.h"

#include <stdexcept>

ReadCashWithdrawalDetailsState::ReadCashWithdrawalDetailsState(ATM* atm) : atm(atm) {}

bool ReadCashWithdrawalDetailsState::cancelTransaction() {
    try {
        atm->changeState(new EjectingCardState(atm));
        return true;
    } catch (const std::exception& e) {
        throw std::logic_error("Cannot Cancel Transaction");
    }
}

bool ReadCashWithdrawalDetailsState::readCashWithdrawDetails() {
    bool isWithdrawValid = true;  // stub logic

    if (isWithdrawValid) {
        atm->changeState(new DispensingCashState(atm));
    } else {
        atm->changeState(new EjectingCardState(atm));
    }

    return isWithdrawValid;
}

ATMState ReadCashWithdrawalDetailsState::getState() {
    return ATMState::READ_CASH_WITHDRAW_DETAILS;
}
